test/Test/Database/LSMTree/Internal/RunReaders.hs

```diff
@@ -0,0 +1,338 @@
+deriving stock instance Show (ReadersObs a)
+deriving stock instance Eq   (ReadersObs a)
+
+instance InLockstep ReadersState where
+  data ModelValue ReadersState a where
+    MEntry  :: SerialisedEntry -> ReadersVal SerialisedEntry
+    MKey    :: SerialisedKey -> ReadersVal SerialisedKey
+    MHasMore:: HasMore -> ReadersVal HasMore
+    MInt    :: Int -> ReadersVal Int
+    MUnit   :: () -> ReadersVal ()
+    MEither :: Either (ReadersVal a) (ReadersVal b) -> ReadersVal (Either a b)
+    MTuple2 :: (ReadersVal a, ReadersVal b)         -> ReadersVal (a, b)
+    MTuple3 :: (ReadersVal a, ReadersVal b, ReadersVal c) -> ReadersVal (a, b, c)
+
+  data Observable ReadersState a where
+    OId     :: (Eq a, Show a) => a -> ReadersObs a
+    OEither :: Either (ReadersObs a) (ReadersObs b) -> ReadersObs (Either a b)
+    OTuple2 :: (ReadersObs a, ReadersObs b) -> ReadersObs (a, b)
+    OTuple3 :: (ReadersObs a, ReadersObs b, ReadersObs c) -> ReadersObs (a, b, c)
+
+  observeModel :: ReadersVal a -> ReadersObs a
+  observeModel = \case
+    MEntry e -> OId e
+    MKey k -> OId k
+    MHasMore h -> OId h
+    MInt n -> OId n
+    MUnit () -> OId ()
+    MEither x -> OEither $ bimap observeModel observeModel x
+    MTuple2 x -> OTuple2 $ bimap observeModel observeModel x
+    MTuple3 x -> OTuple3 $ trimap observeModel observeModel observeModel x
+
+  modelNextState :: forall a.
+       LockstepAction ReadersState a
+    -> ModelLookUp ReadersState
+    -> ReadersState
+    -> (ReadersVal a, ReadersState)
+  modelNextState action lookUp (ReadersState mock) =
+      ReadersState <$> runMock lookUp action mock
+
+  usedVars :: LockstepAction ReadersState a -> [AnyGVar (ModelOp ReadersState)]
+  usedVars = const []
+
+  arbitraryWithVars ::
+       ModelFindVariables ReadersState
+    -> ReadersState
+    -> Gen (Any (LockstepAction ReadersState))
+  arbitraryWithVars _ (ReadersState mock)
+    | isEmpty mock =
+        -- It's not allowed to keep using a drained RunReaders,
+        -- we can only create a new one.
+        Some . New <$> (vector =<< chooseInt (1, 10))
+    | otherwise =
+        QC.frequency $
+          [ (3, pure (Some PeekKey))
+          , (5, pure (Some (Pop 1)))
+          , (1, Some . Pop <$> chooseInt (1, size mock))  -- drain a significant amount
+          , (1, Some . DropWhileKey <$> arbitrary)  -- most likely nothing to drop
+          ] <>
+          [ (3, pure (Some (DropWhileKey k)))  -- drops at least one key
+          | Right k <- [peekKeyMock mock]
+          ]
+
+  shrinkWithVars ::
+       ModelFindVariables ReadersState
+    -> ReadersState
+    -> LockstepAction ReadersState a
+    -> [Any (LockstepAction ReadersState)]
+  shrinkWithVars _ _ = \case
+      New wbs -> Some . New <$> shrink wbs
+      Pop n   -> Some . Pop <$> shrink n
+      _       -> []
+
+  -- TODO: tag, e.g.
+  -- - fully draining the Readers
+  -- - DropWhileKey dropping multiple entries
+  tagStep _ _ _ = []
```

Comment:
I think it's important to add this in a follow-up PR, because otherwise I'm not sure we're exercising some scenarios, like draining readers or `DropWhileKey` dropping anything

Reply:
I added it now, should cover the most interesting cases.

```
    prop_lockstep: OK (8.61s)
      +++ OK, passed 100 tests.
      
      Action polarity (3036 in total):
      100.00% +
      
      Actions (3036 in total):
      45.32% +Pop
      28.89% +DropWhileKey
      20.95% +PeekKey
       4.84% +New
      
      Tags (1772 in total):
      32.67% NewEntriesKeyDuplicates 2 <= n < 4
      28.95% DropWhileKeyDropped 1 <= n < 2
      13.49% DropWhileKeyDropped n == 0
       6.32% NewEntriesKeyDuplicates 4 <= n < 8
       5.47% DropWhileKeyDropped 2 <= n < 4
       5.14% NewEntries 100 <= n < 1000
       3.22% ReadersFullyDrained
       2.99% NewEntries 10 <= n < 100
       1.58% DropWhileKeyDropped 4 <= n < 8
       0.17% NewEntries 1 <= n < 10
```

Reply:
I could tweak it to drain the readers more frequently, but at least it happens.

Reply:
One action you could add is something like `Drain`, which pops until drained. Regardless, this tagging looks like a step in the right direction!

Reply:
I wanted to make sure the last few entries are still checked properly (including looking at the results), as there could be edge cases there.

I guess for `Pop n`, which is currently generated with `n` uniformly between 1 and the number of entries, we could have a special case generating `n` just slightly less than the number of entries. I just tried that, brought the number of `ReadersFullyDrained` from 3% to 7% (or ~47 to ~120 in total numbers), but it also increases the test run time from 7s to 13s, as we are running through a lot more runs that get created.

Reply:
I increased the frequency of all the other actions a little, might be a reasonable tradeoff:

```
    prop_lockstep: OK (9.86s)
      +++ OK, passed 100 tests.
      
      Action polarity (2344 in total):
      100.00% +
      
      Actions (2344 in total):
      44.37% +Pop
      24.23% +PeekKey
      23.85% +DropWhileKey
       7.55% +New
      
      Tags (1536 in total):
      38.54% NewEntriesKeyDuplicates 2 <= n < 4
      23.31% DropWhileKeyDropped 1 <= n < 2
       7.62% NewEntriesKeyDuplicates 4 <= n < 8
       6.71% NewEntries 100 <= n < 1000
       6.64% DropWhileKeyDropped n == 0
       5.73% ReadersFullyDrained
       4.62% DropWhileKeyDropped 2 <= n < 4
       4.36% NewEntries 10 <= n < 100
       1.82% DropWhileKeyDropped 4 <= n < 8
       0.46% NewEntries 1 <= n < 10
       0.20% NewEntriesKeyDuplicates 8 <= n < 16
```